buildcache/build_cache.go

```diff
@@ -0,0 +1,37 @@
+// This file is part of arduino-cli.
+//
+// Copyright 2020 ARDUINO SA (http://www.arduino.cc/)
+//
+// This software is released under the GNU General Public License version 3,
+// which covers the main part of arduino-cli.
+// The terms of this license can be found at:
+// https://www.gnu.org/licenses/gpl-3.0.en.html
+//
+// You can be released from the requirements of the above licenses by purchasing
+// a commercial license. Buying such a license is mandatory if you want to
+// modify or otherwise use the software for commercial activities involving the
+// Arduino software without disclosing the source code of your own applications.
+// To purchase a commercial license, send an email to license@arduino.cc.
+
+package buildcache
+
+import (
+	"time"
+
+	"github.com/arduino/go-paths-helper"
+)
+
+// Used registers the .last-used time in the directory
+func Used(dir *paths.Path) error {
```

Comment:
Ok, I'm not 100% sure of this API, actually, the thing that bothers me much is that `Used` requires a sub-dir of the `basePath`, instead, `Purge` requires directly the `basePath`.

Also `Used` doesn't tell us much about the function, I like much more `GetOrCreate` (that you used inside the `buildCache` object). Also returning the joined path may help to avoid some path mangling on the call side.
 
So I'd like to ask you to try this API:

```go
// GetOrCreate retrieves or create the cache directory for the given key inside basePath.
// If the cache already exists the lifetime of the cache is extended.
func GetOrCreate(basePath *paths.Path, key string) (*paths.Path, error)

// Purge removes all cache directories within baseDir that have expired
// To know how long ago a directory has been last used
// it checks into the .last-used file.
func Purge(baseDir *paths.Path, ttl time.Duration)
```

Reply:
My suggestion was to also return the path to the cache directory.

```go
func GetOrCreate(basePath *paths.Path, key string) (*paths.Path, error)
```

otherwise it forces you to make chains of `.Parent().Parent()` or `.Joins` on the caller side.
BTW I see now that to achieve it a bit more refactoring is required on the `legacy` package, to make it properly, let's talk about it tomorrow.